### PR TITLE
fix: [CDS-78124]: change workflow id and name in case of rollback if tags mismatch

### PR DESCRIPTION
**Pull request by TarunUba1**
## Describe your changes
Handling case when in case of rollback we show wrong workflow against the artifact.

##Testing
Screenshot 2023-09-29 at 7 58 44 PM
Screenshot 2023-09-29 at 7 57 45 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53564)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java [hunk 1]
```diff
@@ -665,15 +667,13 @@ private void processK8sPodsInstances(ContainerInfrastructureMapping containerInf
...
-      deploymentSummary = deploymentSummaryMap.get(containerMetadata);
+    if (deploymentSummary != null) {
       SetView<String> instancesToBeUpdated = Sets.intersection(currentPodsMap.keySet(), dbPodMap.keySet());
-
-      for (String podName : instancesToBeUpdated) {
-        Instance instanceToBeUpdated = dbPodMap.get(podName);
-        K8sPod k8sPod = currentPodsMap.get(podName);
-        String deploymentWorkflowName = Optional.ofNullable(deploymentSummary.getWorkflowExecutionName()).orElse("");
-        if (deploymentWorkflowName.equals(instanceToBeUpdated.getLastWorkflowExecutionName())) {
+      if (deploymentSummary.getK8sDeploymentKey() != null
```
Comment (abosii-harness) [inline 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java:672]: Why we're doing this check here? We should do this check  before retrieving getDeploymentSummaryForInstanceCreation. If releaseNumber will not be present you'll still end up in retrieving the wrong deploymentSummary for newly created instances

Comment (TarunUba1) [reply 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java:675]: added this check for initial deployment summary as well

**Comment by TarunUba1** [comment]
trigger uts

**Comment by TarunUba1** [comment]
trigger ut0, ut3, ut5, ut6

File: 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java [hunk 2]
```diff
@@ -616,6 +617,10 @@ private void processK8sPodsInstances(ContainerInfrastructureMapping containerInf
     }
 
     DeploymentSummary deploymentSummary = deploymentSummaryMap.get(containerMetadata);
+    if (deploymentSummary != null && deploymentSummary.getK8sDeploymentKey() != null
+        && deploymentSummary.getK8sDeploymentKey().getReleaseNumber() != null) {
+      getDeploymentSummaryForInstanceCreation(deploymentSummaryMap.get(containerMetadata), rollback);
```
Comment (abosii-harness) [inline 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java:622]: please make it more clear, use if else

File: 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java (hunk 1 above)
Comment (abosii-harness) [reply 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java:675]: Do we still need to do this check? What is the point of it?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java (hunk 1 above)
Comment (TarunUba1) [reply 400-rest/src/main/java/software/wings/service/impl/instance/ContainerInstanceHandler.java:675]: We still need that condition cause like 
canary rollback :- then it should not change the workflow if workflow name different from existing instance workflow while updating
so i am having a check when release number is not present to not update when workflow name different

**Comment by TarunUba1** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

**Review by GoelPratyush99 — APPROVED** [review]

**Comment by TarunUba1** [comment]
trigger messagecheck, review

**Pull request merged**